API: Error while inviting user to workspace

**Describe the bug**

When we try to invite a user to a workspace, and we try to assign them a role ID that doesn't exist, instead of returning a 404 with proper error message, the API throws a 500 error.

**To Reproduce**
Steps to reproduce the behavior:
1. Create a workspace
2. Try to invite another user to the workspace and in `roleIds` array, supply an ID that doesn't exist
3. Encounter the error

**Expected behavior**

If a workspace role provided in the `roleIds` array doesn't exist, we want to return a 404 specifying that the particular role doesn't exist.

**Screenshots**
- Postman request:
![image](https://github.com/user-attachments/assets/5ab63ad4-8f55-410f-97ef-6ae5b1e3540a)
- Response:
![image](https://github.com/user-attachments/assets/d6d62926-f8be-464e-9c68-2b770b38ac9e)
- Server logs:
```
api:dev: [ERROR] 2024-07-30 14:21:38 - 
api:dev: Invalid `this.prisma.workspaceMember.create()` invocation in
api:dev: /home/raj/oss/keyshade.xyz/keyshade-xyz/apps/api/src/workspace/service/workspace.service.ts:983:60
api:dev: 
api:dev:   980 }
api:dev:   981 
api:dev:   982 // Create the workspace membership
api:dev: → 983 const createMembership = this.prisma.workspaceMember.create(
api:dev: An operation failed because it depends on one or more records that were required but not found. No 'WorkspaceRole' record(s) (needed to inline the relation on 'WorkspaceMemberRoleAssociation' record(s)) was found for a nested connect on one-to-many relation 'WorkspaceMemberRoleAssociationToWorkspaceRole'.
```

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser: Postman
 - Version: 2.3.0

**Additional context**

We also doubt that the same error will be present in `Update member role` endpoint aswell.

--- a/apps/api/src/workspace/service/workspace.service.ts
+++ b/apps/api/src/workspace/service/workspace.service.ts
@@ -414,18 +414,9 @@ export class WorkspaceService {
         }
       })
 
-    //Check if the roleIds are valid
-    const roles = await this.prisma.workspaceRole.findMany({
-      where: {
-        id: {
-          in: roleIds
-        },
-        workspaceId: workspace.id
-      }
-    })
-
-    const invalidRoles = roleIds.filter(
-      (id) => !roles.map((role) => role.id).includes(id)
+    const invalidRoles = await this.findInvalidWorkspaceRoles(
+      workspace.id,
+      roleIds
     )
 
     if (invalidRoles.length > 0) {
@@ -1000,18 +991,9 @@ export class WorkspaceService {
         )
       }
 
-      //Check if the member has valid roles
-      const roles = await this.prisma.workspaceRole.findMany({
-        where: {
-          id: {
-            in: member.roleIds
-          },
-          workspaceId: workspace.id
-        }
-      })
-
-      const invalidRoles = member.roleIds.filter(
-        (id) => !roles.map((role) => role.id).includes(id)
+      const invalidRoles = await this.findInvalidWorkspaceRoles(
+        workspace.id,
+        member.roleIds
       )
 
       if (invalidRoles.length > 0) {
@@ -1083,6 +1065,26 @@ export class WorkspaceService {
 
       this.log.debug(`Added user ${memberUser} to workspace ${workspace.name}.`)
     }
+  }
+
+  private async findInvalidWorkspaceRoles(
+    workspaceId: string,
+    roleIds: string[]
+  ) {
+    const roles = await this.prisma.workspaceRole.findMany({
+      where: {
+        id: {
+          in: roleIds
+        },
+        workspaceId: workspaceId
+      }
+    })
+
+    const roleIdSet = new Set(roles.map((role) => role.id))
+
+    const invalidRoles = roleIds.filter((id) => !roleIdSet.has(id))
+
+    return invalidRoles
   }
 
   private async memberExistsInWorkspace(

--- a/apps/api/src/workspace/service/workspace.service.ts
+++ b/apps/api/src/workspace/service/workspace.service.ts
@@ -414,6 +414,26 @@ export class WorkspaceService {
         }
       })
 
+    //Check if the roleIds are valid
+    const roles = await this.prisma.workspaceRole.findMany({
+      where: {
+        id: {
+          in: roleIds
+        },
+        workspaceId: workspace.id
+      }
+    })
+
+    const invalidRoles = roleIds.filter(
+      (id) => !roles.map((role) => role.id).includes(id)
+    )
+
+    if (invalidRoles.length > 0) {
+      throw new NotFoundException(
+        `Workspace ${workspace.name} (${workspace.id}) does not have roles ${invalidRoles.join(', ')}`
+      )
+    }
+
     // Create new associations
     const createNewAssociations =
       this.prisma.workspaceMemberRoleAssociation.createMany({
@@ -977,6 +997,26 @@ export class WorkspaceService {
         )
         throw new ConflictException(
           `User ${memberUser.name} (${userId}) is already a member of workspace ${workspace.name} (${workspace.id})`
+        )
+      }
+
+      //Check if the member has valid roles
+      const roles = await this.prisma.workspaceRole.findMany({
+        where: {
+          id: {
+            in: member.roleIds
+          },
+          workspaceId: workspace.id
+        }
+      })
+
+      const invalidRoles = member.roleIds.filter(
+        (id) => !roles.map((role) => role.id).includes(id)
+      )
+
+      if (invalidRoles.length > 0) {
+        throw new NotFoundException(
+          `Workspace ${workspace.name} (${workspace.id}) does not have roles ${invalidRoles.join(', ')}`
         )
       }
 

